### ISSUE TITLE
Option to disable Spaces bar

### ISSUE THREAD

**Issue by BurnyBoi** [vector-im/element-web#18898]
### Your use case

When using a vertical monitor or using Element in a vertical window, the Spaces bar on the left can take up a lot of unnecessary space. There is a way to disable the communities bar, but not this new Spaces one.

There should be a way to hide the bar. Once disabled, Spaces would either need to be accessed some other way or have the bar re-enabled when there is a need to access them.

### Have you considered any alternatives?

Currently Spaces can be kept off and aren't used so for now I'm leaving it off but eventually this will be enabled for all.

**Comment by aaronraimist** [comment]
Related: https://github.com/vector-im/element-web/issues/8733

**Comment by gdvine** [comment]
Related: https://github.com/vector-im/element-web/issues/19191

**Comment by cbs228** [comment]
What about optionally relocating the spaces bar next to the room directory button? Mockup:

![mockup_move_spaces_button](https://user-images.githubusercontent.com/3526918/135701697-b1e7bfe0-162b-479d-ad8d-4d42a500aac2.png)

Clicking the button would reveal available spaces.

If the option to "minimize spaces" is selected, it should probably imply that "`Show all rooms in Home`" is enabled.

**Comment by BurnyBoi** [comment]
> > Currently Spaces can be kept off and aren't used so for now I'm leaving it off but eventually this will be enabled for all.
> 
> Yeah, I had a way to hide it (the fixed-width bar with 2 icons on the left) but this option is said to be deleted in the near future.

For Element Web, you can also use the uBlock Origin extension to add a rule to hide the bar. This works well for now and spaces can still be accessed by links.

**Comment by bdefore** [comment]
Until there's a preference for this, there's a hacky workaround for web users with uBlock Origin extension on. Right click in the Spaces area, select 'Block Element', then click 'Create' (to create a new record for a blocked DOM element). You may need to do it twice.

**Comment by konomikitten** [comment]
I just updated to Element Desktop v1.9.8 and now it seems the User menu has been moved into the spaces column, a change like that seems to indicate to me that having this column hidden in the future would now be more difficult because the spaces column now contains something that has nothing to do with spaces itself.

I personally do not find spaces useful, I find the extra amount of UI navigation just to read the rooms contained within spaces tedious and annoying. I'd like to see a way to disable spaces properly or at least make them far more useful than they are now.

**Comment by aaronraimist** [comment]
@konomikitten It looks like that has been taken care of in https://github.com/matrix-org/matrix-react-sdk/pull/7258

**Comment by 4nd3r** [comment]
When this change finally lands, please make it configurable with `UIFeature.spaces` or some similar UI feature flags. Thanks.

**Comment by t3chguy** [comment]
All settings (vast majority) are configurable within the settingDefaults key

**Comment by meichthys** [comment]
> All settings (vast majority) are configurable within the settingDefaults key

Is there a reference somewhere that shows the settings that are configurable, and their respective values?

**Comment by RenWal** [comment]
@meichthys  https://github.com/vector-im/element-web/blob/develop/docs/config.md#ui-features

**Comment by meichthys** [comment]
> @meichthys  https://github.com/vector-im/element-web/blob/develop/docs/config.md#ui-features

Thanks, I did find that although I was mostly looking for the options that are configurable under the 'settingDefaults' Setting described in number 20 on that document.

**Comment by t3chguy** [comment]
@meichthys https://github.com/matrix-org/matrix-react-sdk/blob/develop/src/settings/Settings.tsx#L365-L885

**Comment by meichthys** [comment]
> @meichthys https://github.com/matrix-org/matrix-react-sdk/blob/develop/src/settings/Settings.tsx#L365-L885

Perfect, thanks!

**Comment by BurnyBoi** [comment]
FYI, develop.element.io has now disabled communities, and this change was never merged in, so soon there will be no choice to hide the bar soon. Can this be merged before removing communities goes live? Those of us using vertical monitors would appreciate this!

**Comment by t3chguy** [comment]
Given the designer responsible for this area is off at the moment that will unfortunately not happen.

**Comment by MonkeyBars3k** [comment]
"Settings > Prefs > Display Communities instead of Spaces" has been removed in today's release, so now I don't know of a way to hide this large sidebar that unnecessarily takes up space (pun intended) for users not interested in the feature.

This is a UX downgrade for us. Does anyone know of a way to hide this thing on the desktop client?

**Comment by ostwilkens** [comment]
Today the Spaces bar popped up again, and I can't find a way to disable it. 
My current workaround is to use uBlock to hide the div. 

Element is for me, since long ago, feature complete. What I am now missing is to disable features I don't need. Spaces and markdown come to mind.

**Comment by meichthys** [comment]
> This is a UX downgrade for us. Does anyone know of a way to hide this thing on the desktop client?

It's not ideal and won't work for large deployments, but a quick hack is:
`Ctrl - Shift - I` then search for `mx_SpacePanel collapsed` in the `Elements` pane then in the `Styles` sub-pane scroll to the bottom and click into the center box shown below and modify the first number to be `0`. 
This seems to survive a restart of the app and also preserves the user provile badge.
![image](https://user-images.githubusercontent.com/10717998/163006038-9124c78b-6a4c-4d37-bf47-e036c1d05f82.png)

**Comment by thany** [comment]
Try this bit so you won't loose the profile button that also inexplicably moved to the new monstrosity bar:

```css
.mx_LeftPanel_wrapper {
  flex-direction: columns;
}
.mx_LeftPanel_wrapper--user {
  flex-grow: 1;
}
```

As said, it doesn't survive a restart if you apply it through devtools (and frankly, regular users shouldn't have to be put through this, but instead new releases should be put through better testing procedures).

If you can find how to do it permanantly, please let us know.

**Comment by wezm** [comment]
I use Element on the web and use custom CSS to hide the sidebar permanently with [Stylus](https://github.com/openstyles/stylus). My userstyles for the Element instance look like this:

```css
*[aria-label="Spaces"] {
    display: none;
}
```

**Comment by gcp** [comment]
That has the same problem in that it hides the profile button which you now need to access the Element settings. (If that would have worked, the Spaces sidebar could've just been zapped by uBlock Origin).

>Those of us using vertical monitors would appreciate this!

If you're using multiple windows side by side this change is also a large usability problem.

**Comment by BurnyBoi** [comment]
For the time being, I've been using this CSS to make the spaces bar a bit more compact:

```
:root {
    --element-space-panel-collapsed-width: .5em;
    --element-space-panel-dim-left: calc(4.1em);
}

.mx_RoomHeader_buttons .mx_RoomHeader_videoCallButton,
.mx_RoomHeader_buttons .mx_RoomHeader_voiceCallButton,
.mx_MessageComposer_row .mx_MessageComposer_voiceMessage,
.mx_MessageComposer_row .mx_MessageComposer_stickers
{ display: none; }

.mx_SpaceCreateMenu_wrapper .mx_ContextualMenu_background {
    left: var(--element-space-panel-dim-left)!important;
}

#matrixchat .mx_SpacePanel.collapsed .mx_SpaceTreeLevel {
    padding: 0!important
}

#matrixchat .mx_SpacePanel.collapsed .mx_SpaceItem .mx_SpaceButton {
    padding: .2em 0!important
}

#matrixchat .mx_SpacePanel.collapsed .mx_SpaceButton,
#matrixchat .mx_SpacePanel.collapsed .mx_SpacePanel_toggleCollapse {
    margin-left: calc(16px - calc(var(--element-space-panel-collapsed-width) + .1em))!important;
    margin-right: var(--element-space-panel-collapsed-width);
}

#matrixchat .mx_SpacePanel.collapsed .mx_UserMenu {
    margin: 10px!important;
}

```

**Comment by odiferousmint** [comment]
It is very easy to get rid of stuff you do not need once you open the Developer Tools. The issue is making it persistent. Any ideas to achieve this on Linux?

**Comment by wezm** [comment]
> It is very easy to get rid of stuff you do not need once you open the Developer Tools. The issue is making it persistent. Any ideas to achieve this on Linux? Most optimal way is preferred, i.e. without third-party application, if possible.

I use Stylus on Linux (see [my comment above](https://github.com/vector-im/element-web/issues/18898#issuecomment-1098577021)

**Comment by odiferousmint** [comment]
> > It is very easy to get rid of stuff you do not need once you open the Developer Tools. The issue is making it persistent. Any ideas to achieve this on Linux? Most optimal way is preferred, i.e. without third-party application, if possible.
> 
> I use Stylus on Linux (see [my comment above](https://github.com/vector-im/element-web/issues/18898#issuecomment-1098577021)

I saw that. Is it supposed to work with the Electron app (Element Desktop)? I apologize, I forgot to mention that I was talking about it. You can open Developer Tools inside the Element app as well, and it is easy to get rid of what you do not need, but making it persistent is less so.

**Comment by wezm** [comment]
> Is it supposed to work with the Electron app (Element Desktop)?

No, it's for web browser (Firefox in my case).

**Comment by odiferousmint** [comment]
> > Is it supposed to work with the Electron app (Element Desktop)?
> 
> No, it's for web browser (Firefox in my case).

Yeah, that is what I thought. I do use a browser extension for such purposes so that would not be much of an issue, but the desktop app is. :/ I hope that someone knows a way.

**Comment by thany** [comment]
We shouldn't be hacking away like that. If users demand change, change must occur. Simple as that.

**Comment by SomeChud** [comment]
Bomb detonation in 3 days if this is not done. Allah is waiting!

**Comment by odiferousmint** [comment]
> We shouldn't be hacking away like that. If users demand change, change must occur. Simple as that.

Agreed. It should not even be much of a big deal to keep it as it ORIGINALLY was, or at least revert the commit that removed the setting which at the very least gave us the ability to have it the old way.

I personally do not care about Spaces. I have no clue why I should care about it. I care about People, and Rooms. If you add the ability to switch between accounts, then please, do not use a bar for that either, just let us click on the avatar which would drop down a menu that includes "Switch account" or something.

Additionally, well, this is a feature request, but it would be neat if we had plus two options: 1) completely hide the original / initial bar as well,  2) make it appear only when the cursor is nearby the edge.

**Comment by t3chguy** [comment]
Reverting that commit would result in bringing back communities and all the tech debt associated, which simply won't happen.

**Comment by odiferousmint** [comment]
> Reverting that commit would result in bringing back communities and all the tech debt associated, which simply won't happen.

I suppose that is a huge commit or a set of commits then, right? What exactly is the difference between Communities and Spaces, and why the latter over the former? Any pointers?

**Comment by t3chguy** [comment]
Communities:
+ Centralised
+ Not part of the Matrix spec
+ Flat structure
+ Buggy
+ Doesn't reuse existing Matrix data structures

Spaces:
+ Opposite of all of those

**Comment by odiferousmint** [comment]
> Communities:
> 
> * Centralised
> * Not part of the Matrix spec
> * Flat structure
> * Buggy
> * Doesn't reuse existing Matrix data structures
> 
> Spaces:
> 
> * Opposite of all of those
> 

I see. I am not sure what you mean by flat structure, but neither of those should be an issue to make Spaces act / seem / feel like Communities, right? We just want one bar at the left side (if not customizable) that can be re-sized and collapsed similarly to how the Communities one could. That is all possible without undermining any of those points, is this correct?

**Comment by t3chguy** [comment]
Spaces can contain other Spaces

The communities bar could not be resized, it was a fixed size to the left of the room list.

**Comment by BurnyBoi** [comment]
True, that's why I always enabled the option to hide it. I just want that same option to hide this new bar.

**Comment by thany** [comment]
> Reverting that commit would result in bringing back communities and all the tech debt associated, which simply won't happen.

Shouldn't have put everything in a big giant commit then, ey? Something broke that worked before. That means a revert has to be performed straight away. End of story. This is just part of professional software development in general.

As a user, I don't care about tech debt. It's not part of the Spaces feature in the first place, but is (or should be) part of a different issue, with different commits associated to it. And to be brutally honest, as a user, I don't care about Spaces either. No amount of advantages anyone lists, will convince me, because I simply do not use it.

Also bear in mind there's no shame in having to revert functionality. It happens to the best of us. You should see it as an oppertunity to refine what you thought was already perfect, and to align it better with users' demands.

**Comment by t3chguy** [comment]
> Shouldn't have put everything in a big giant commit then, ey? Something broke that worked before. That means a revert has to be performed straight away. End of story. This is just part of professional software development in general.

@thany that's not how it works. You can't remove all the supporting code for communities yet keep the communities bar.
Synapse has stopped supporting communities also, so you'd be met with lots of errors trying to keep that code around.

**Comment by thany** [comment]
That *is* how it works, but in this *particular* case, a revert is no longer possible due to upstream API limitations, is what you're saying.

All the more reason to give this problem some elevated priority!

**Comment by t3chguy** [comment]
> That is how it works, but in this particular case, a revert is no longer possible due to upstream API limitations, is what you're saying.



No. If you have feature X which depends on feature Y and you remove both X then feature Y, you can't just bring back feature X regardless of if they were removed in one commit or not. The supporting code will have gone and all you'll have is compilation errors.

**Comment by BurnyBoi** [comment]
I feel like the wrong terminology has been used just now. No one's asking for the communities bar to be brought back. This doesn't require any code to be "reverted." We're just asking for the new spaces bar to be improved so it functions like the communities bar did (in that you can always toggle it off). The people have already spoken and this is one of the most upvoted open issues right now with plenty of good reasons why this feature is 100% necessary, so it just seems silly at this point to not wrap this up since the work was already mostly done [here](https://github.com/matrix-org/matrix-react-sdk/pull/7128) and just needs to be merged into the latest SDK, then tweaked a bit more to account for a few more changes to the left bar since then.

**Comment by odiferousmint** [comment]
> > so it just seems silly at this point to not wrap this up since the work was already mostly done [matrix-org/matrix-react-sdk#7128](https://github.com/matrix-org/matrix-react-sdk/pull/7128) and just needs to be merged into the latest SDK, then tweaked a bit more to account for a few more changes to the left bar since then.
> 
> No point wasting development cycles until product & design agree to what the UX will be like for users who disable it when they interact with spaces

Perhaps there can be options to:

1. Completely disable it
2. Arrow icon in the middle horizontally that would display the bar
3. Increased border size to know there is a bar collapsed
4. Display the bar when the mouse gets close to the edge where the bar is

I am pretty sure there are other ways to go about it as well. I actually forgot how Spaces worked, but perhaps anything similar may do the job.

**Comment by bdefore** [comment]
There's really not a window for the community to provide a solution in code here. The Spaces feature introduced a design problem and unfortunately shipped without sufficient consideration of use case of those who... don't care about Spaces. Let's complete the loop. Please raise the importance with that design team to consider the minimal use case. It's a messaging application first.

**Comment by t3chguy** [comment]
> Completely disable it

Yup this is the way the PR goes, the issue is product won't land a feature like that as then you have no way to leave/manage spaces you are already in, there's a wider reaching effect here.

**Comment by weeman1337** [comment]
Questions to be answered:
- Users will loose room discovery in spaces
- Some room permissions depend on spaces

I think because of that users should still be able to join and explore spaces.

Currently some mock-ups exist here: https://www.figma.com/file/QTxqmDYaQiSU4I2KYHz35T/%5BSpaces%5D-Web---hide-spaces?node-id=3%3A3024

**Some ideas that came into my mind:**

Disabling spaces should enable „show everything in home space“

Instead of a magical switcher we could integrate spaces flat into the people/room list. Clicking on it could lead to the space home. So that users can leave spaces / discover the rooms from there.

[spaces in list](https://user-images.githubusercontent.com/6216686/177749580-eb88aae1-535a-4bfc-b7fe-f63e796d338e.png)

**Comment by t3chguy** [comment]
> Questions to be answered:

Also what to do with the spaces the user is already in, there should be a way to leave them given they'll be wasting /sync bandwidth

**Comment by BurnyBoi** [comment]
>This option should be available only when the end user doesn't join any space

Having this as an option for an Element instance seems beneficial, but ultimately, this thread is still about making the spaces bar always toggleable, regardless whether the user has already joined a space. As far as UI, there's been suggestions of a mini menu to still access spaces when the bar is hidden, as well as having a way to just quickly toggle the bar back on when needed.

**Comment by giomfo** [comment]
@gaelledel I think we may re-open https://github.com/vector-im/element-web/issues/22765 to handle separately these multiple requests about Spaces

**Comment by rufuskahler** [comment]
Removing 'Needs-Design' as will be resolved through EX work.

**Comment by t3chguy** [comment]
@rufuskahler removing X-Needs-Design means anyone can just start it with assets already provided in the issue, (e.g. community) - that does not appear to be the case, maybe you meant to remove `X-Needs-Product` - as without knowing how design expects the interaction to look this will be impossible to contribute.

**Comment by thany** [comment]
The spaces bar can be removed as-is. What additional assets could you possibly need to *remove* something?

**Comment by t3chguy** [comment]
@thany the product team doesn't want it removed, they want the functionality to be available elsewhere, so the UX to switch behaviours and the alternative behaviour need designs.

**Comment by thany** [comment]
A checkbox should suffice.

**Comment by t3chguy** [comment]
@thany that doesn't explain how a user can manage spaces they are in without the space panel

**Comment by t3chguy** [comment]
Simply hiding spaces & invites is very unlikely to fly, they'll need exposed via alternate means, like a dropdown more akin to EX

**Comment by thany** [comment]
Spaces don't *need* to be exposed at all. If a user hides Spaces, it's for a reason. I can only really speak for myself, so when I hide spaces, it's because I don't use them and I don't want them taking up any space (especially not when they expand themselves at the whim of increasing window size, which I have to do every time Element starts because it's not able to remember its own size properly).

Users who do need spaces, can keep them visible. I really, honestly, genuinely don't see the problem here. Just hide the bloody thing, or don't. How hard can it be?

Either way, in my case, what would be my advantage of still having access to spaces?

**Comment by t3chguy** [comment]
> Spaces don't need to be exposed at all

That's up to the Element Product team

**Comment by thany** [comment]
No, it should be up to the users.

**Comment by t3chguy** [comment]
Indirectly it is, but not up to the vocal users, but up to the wider userbase. The vocal minority often do not match the average.

**Comment by thany** [comment]
> Either way, in my case, what would be my advantage of still having access to spaces?

Maybe they can also answer this, because I can't think of anything.

**Comment by jellykells** [comment]
Can someone please point me to the overwhelming user feedback that says "please don't let anyone hide the Spaces bar I want every user to have to see it"

**Comment by t3chguy** [comment]
I suggest waiting to see what the Product & Design teams have planned in this space rather than demanding a given solution. Users present issues & feedback, they don't get to drive the choices and implementation. The product is free. You're always welcome to fork it and make it work exactly as you hope and dream.

**Comment by jellykells** [comment]
So users don't really drive the choices made by the Element team. Well, thanks for being honest at least.

**Comment by t3chguy** [comment]
You can say "the contrast" is bad, but not pick our colour palette. Just like here you can say you want a way to get rid of the stacked space panel bar taking up your valuable space, doesn't mean you can decide that there'll also be no way to manage spaces alongside that. If you want your own software designed just like you want it, try freelancer or fiverr.

**Comment by MonkeyBars3k** [comment]
I have tried making Spaces and been reading about them and still can't wrap my head around them _at all_. Is there a tutorial somewhere out there? I find the explanations on the blog extremely confusing. At its core, what even IS a "Space"??

**Comment by jellykells** [comment]
> You can say "the contrast is bad", but not pick our colour palette

Not a great example, since the Element team has actually gone in the right direction regarding theming and given users the ability to pick whatever color palette they want. 

Similarly, I think it is reasonable to give users who aren't interested in seeing or managing spaces a way to hid the interface intended to facilitate that.

**Comment by t3chguy** [comment]
At the end of the day a space is a room, if you're in a space and something happening in that space makes your push rules trigger, you'll need to see it otherwise you'll be getting pings/notifications without an ability to see them. You can't disable push rules for spaces in the spec today. Equally a space could create a lot of /sync bloat due to being filled with state events, so you would want to retain a way to leave it without needing to reconfigure your client. Again can't filter out spaces in /sync with the filters provided by the spec.

**Comment by jellykells** [comment]
I think /sync returning unnecessary data if the UI were disabled is a separate issue that shouldn't necessarily block this feature request, and I wonder how many users would even be aware that it was occurring.

As far as push rules go, I think the only notifications triggered by default for spaces is invites? I can see why you wouldn't want users getting an invite notification to a space and then not being able to view it, that seems like a legitimate concern. On the other hand, for users not interested enough in using spaces to have the UI visible, it seems minor enough that the Space bar toggle could be added as a Labs feature until someone might be willing to put in a spec PR to improve the /sync filters.

On that train of thought, would a PR implementing such a Labs feature even be considered for merge, or is this really "fork it" territory?

**Comment by t3chguy** [comment]
> On that train of thought, would a PR implementing such a Labs feature even be considered for merge, or is this really "fork it" territory?

It is unlikely to be accepted given the product team has it in their sights, unless the author was willing to shape the PR into whatever the product & design teams had in mind

**Comment by jellykells** [comment]
Personally, I find Spaces somewhat useful (even if we don't have user flair back yet!), but the large and ever-present UI obnoxious. So for me this feature request  is more like "let me hide it while you guys brainstorm something better" rather than a final solution, hence considering a Labs flag.

But I suppose if there's resistance to even adding a stop-gap solution, that's a dead end.

**Comment by t3chguy** [comment]
We're just averse to adding yet more unmaintained labs flags which are poorly covered by tests and increase tech debt. If someone wanted to revive https://github.com/matrix-org/matrix-react-sdk/pull/7128 and leverage our Netlify builds to keep it hosted that could be an easy interim solution

**Comment by sallyhaj** [comment]
That's a good idea, whether to hide Space side bar or remove it completely.. It takes lot of space without an option to minimize/hide it.
Thank you.

![Image](https://github.com/user-attachments/assets/d5303a89-5d4c-47ec-943c-16798328e24c)
